### PR TITLE
chore(node): Pass around MyNodeState to avoid holding locks

### PR DESCRIPTION
**Pull request by joshuef**
For longer running message handling, we now pass around the inital MyNodeState. This avoids a tonnnn of read locks and therefore hopefully prevents holding up write and reads needlessly.

### CODE REVIEW

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

**Comment by joshuef** [comment]
bors try

File: sn_node/src/node/mod.rs [hunk 1]
```diff
@@ -148,7 +149,47 @@ mod core {
...
+        pub(crate) event_sender: EventSender,
+        pub(crate) joins_allowed: bool,
+        pub(crate) membership: Option<Membership>,
+    }
+
     impl MyNode {
+        /// Generate a read only snapshot of a subset of the node's state
+        /// Useful for longer running processes to avoid having to acquire
+        /// read locks eg
+        pub(crate) fn get_snapshot(&self) -> MyNodeSnapshot {
```
Comment (b-zee) [inline sn_node/src/node/mod.rs:174]: Rust convention is to not use `get_` prefixes.

File: sn_node/src/node/api.rs [hunk 2]
```diff
@@ -103,8 +105,12 @@ impl MyNode {
     }
 
     /// Returns the SAP of the section matching the name.
-    pub(crate) fn matching_section(&self, name: &XorName) -> Result<SectionAuthorityProvider> {
-        self.network_knowledge
+    pub(crate) fn matching_section(
```
Comment (b-zee) [inline sn_node/src/node/api.rs:108]: Perhaps better to move this to the `impl MyNodeSnapshot`?

**Review by b-zee — COMMENTED** [review]
Generally I have mixed feelings as it doesn't help readability of code. It seems like a sledge hammer approach, because I assume it solves problems, but it's not exactly sure where the exact bottleneck was.
Overall I think it's an improvement for performance, but I can't verify.

**Comment by joshuef** [comment]
Generally what I was seeing was some longer than desireable waits on `locks`, that could be `read` or `write` locks. 

So yeh, this is a sledgehammer to reduce that for sure. In general I think it clears up some misplaced mutability, and removes potential for deadlocks if we're holding them for longer than desired. (And we had a bunch over `await` calls).

I think  off the back of this we could look to pull out a bunch of functionality from the node into some more testable functions eg. (everything that is now `MyNode::<X>`. Which could take distinct arguments as opposed to the whole state. (That was much simpler for now though).

Here I feel it's also simpler to reason about the code and where we do need to mutate some state. It's not as often as you think and removing the `self` code I think helps clarify that...

**Comment by joshuef** [comment]
bors try

File: sn_node/src/node/mod.rs [hunk 3]
```diff
@@ -254,30 +309,30 @@ mod core {
         ////////////////////////////////////////////////////////////////////////////
 
         /// Generates a random AE probe for _anywhere_ on the network.
-        pub(crate) fn generate_probe_msg(&self) -> Result<Cmd> {
+        pub(crate) fn generate_probe_msg(snapshot: &MyNodeSnapshot) -> Result<Cmd> {
```
Comment (bochaco) [inline sn_node/src/node/mod.rs:312]: It sounds like all APIs like this one which work on a snapshot shall now be part of the `MyNodeSnapshot` impl instead?

File: sn_node/src/node/mod.rs [hunk 4]
```diff
@@ -723,12 +777,12 @@ mod core {
...
-    #[derive(Clone)]
-    pub(crate) struct StateSnapshot {
-        pub(crate) is_elder: bool,
-        pub(crate) section_key: bls::PublicKey,
-        pub(crate) prefix: Prefix,
-        pub(crate) elders: BTreeSet<XorName>,
-        pub(crate) members: BTreeSet<XorName>,
-    }
+    // #[derive(Clone)]
+    // pub(crate) struct StateSnapshot {
```
Comment (bochaco) [inline sn_node/src/node/mod.rs:781]: I guess we can remove this struct

File: .github/workflows/run_pr_checks.yml [hunk 5]
```diff
@@ -191,7 +191,7 @@ jobs:
 
       - name: Run client tests
         env:
-          RUST_LOG: "sn_client=trace,qp2p=debug"
+          RUST_LOG: "sn_client,sn_interface"
```
[the remark this answers is not in the document]
Comment (joshuef) [reply .github/workflows/run_pr_checks.yml:194]: its not needed, this'll just default to _all_ logs from that lib.

File: sn_node/src/node/mod.rs (hunk 3 above)
Comment (joshuef) [reply sn_node/src/node/mod.rs:312]: Aye they could be. I wasn't sure where to draw the line atm. I think there's a lot more refactoring we could do atop this... but maybe better for follow up PRs?

File: sn_node/src/node/flow_ctrl/dispatcher.rs [hunk 6]
```diff
@@ -100,20 +104,35 @@ impl Dispatcher {
...
-                // drop the write lock
-                drop(node);
-
-                let node = self.node.read().await;
-
-                debug!("Network knowledge was updated: {updated}");
-                node.handle_valid_client_msg(msg_id, msg, auth, origin, send_stream)
+                let mut snapshot = self.node.read().await.snapshot();
+                // block off the write lock to ensure it's dropped
+                {
```
Comment (bochaco) [inline sn_node/src/node/flow_ctrl/dispatcher.rs:109]: Isn't the scope provided by `if there_was_an_update {` below enough for dropping the lock ASAP ?

File: sn_node/src/node/mod.rs (hunk 3 above)
Comment (bochaco) [reply sn_node/src/node/mod.rs:311]: I see, yeah, makes sense

File: sn_node/src/node/flow_ctrl/dispatcher.rs (hunk 6 above)
Comment (joshuef) [reply sn_node/src/node/flow_ctrl/dispatcher.rs:109]: the read lock? 

That should be released at the end of the line I believe: `let mut snapshot = self.node.read().await.snapshot();` , (The snapshot itself is pure cloning, so mutability here doesnt actually affect `MyNode` or need a lock).

Or do you mean something else?

Comment (joshuef) [reply sn_node/src/node/flow_ctrl/dispatcher.rs:109]: ahhhhh with you. yeh, you're totally right! That's a hangover from an earlier changeset I think. I'll clear it up.

File: sn_node/src/node/mod.rs [hunk 7]
```diff
@@ -148,7 +148,59 @@ mod core {
         pub(crate) dysfunction_tracking: DysfunctionDetection,
     }
 
+    #[derive(Clone)]
+    pub(crate) struct MyNodeSnapshot {
+        pub(crate) root_storage_dir: PathBuf,
+        pub(crate) is_elder: bool,
+        pub(crate) is_not_elder: bool,
```
Comment (bochaco) [inline sn_node/src/node/mod.rs:155]: I think we shall remove either `is_elder` or `is_not_elder`

**Review by bochaco — COMMENTED** [review]
As a general comment, perhaps we shall use a diff name than `MyNodeSnapshot`/`snapshot`, I'm thinking perhaps something like `context` as it's the pattern we are actually implementing here...?

**Comment by joshuef** [comment]
bors try

**Review by bochaco — APPROVED** [review]

**Pull request merged**